articles/external_qos_file.md

```diff
@@ -0,0 +1,638 @@
+
+<b id="to_node_name_uniqueness">1</b> Node name uniqueness is not being enforced up to Foxy, though it is assumed. [↩](#back_node_name_uniqueness)
+
+### Nodes
+
+QoS settings apply to entities that are associated with ROS nodes.
+Thus, a top level tag to identify the node should be available in the file format.
+Here are a couple of examples:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        ...
+    </node>
+</qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        ...
+```
+
+### Entity type
+
+It should be possible to distinguish between the different entity types in a QoS profile file.
+For example:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        <publisher topic_name="asd">  <!--Relative name-->
+            ...
+        </publisher>
+        <subscription topic_name="/another_ns/asd">  <!--Absolute name-->
+            ...
+        </subscription>
+    </node>
+</qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        publisher:
+            topic_name: asd
+            qos:
+                ...
+        subscription:
+            topic_name: /another_ns/asd
+            qos:
+                ...
+```
+
+Besides `publisher` and `subscription` tags, `client` and `service` should be allowed.
+
+### QoS profiles IDs (optional)
+
+In the case that a node wants to create two `publishers` on the same topic with different QoS, the above format wouldn't allow to uniquely identify the `publisher`. The same applies to other kinds of entitys.
+
+To solve this issue, the format could support using a profile ID, in addition to the topic name:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        <publisher topic_name="asd" profile_id="my_pub_profile_1">
+            ...
+        </publisher>
+        <publisher topic_name="asd" profile_id="my_pub_profile_2">
+            ...
+        </publisher>
+    </node>
+</qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        publisher:
+            topic_name: asd
+            profile_id: my_pub_profile_1
+            qos:
+                ...
+        publisher:
+            topic_name: asd
+            profile_id: my_pub_profile_2
+            qos:
+                ...
+```
+
+`rcl` API would need to be extended to support profiles IDs.
+Users should only use this mechanism to avoid collisions.
+
+### Default profiles for entities in different nodes
+
+It is common to have several nodes creating the same entity, and the same QoS profile is desired in all of them.
+A default tag can be added to solve this:
+
+```xml
+<qos_profiles>
+    <default>
+        <publisher topic_name="/my_ns/nested_ns/asd">
+            ...
+        </publisher>
+    </default>
+    <node name="my_node" namespace="/my_ns/nested_ns">
```

Comment:
Would it be reasonable to support wildcards in the names?

Reply:
that sounds reasonable to me

Reply:
See https://github.com/ros2/design/pull/296/commits/d22a11cdb6d97417631cb4f5aa7de93f1c761396.